### PR TITLE
Hubs are more fleshed out

### PR DESCRIPTION
**Pull request by davidfowl**

### CODE REVIEW

File: samples/SocketsSample/HubCallerContext.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
+﻿using System.Security.Claims;
+using Microsoft.AspNetCore.Sockets;
+
+namespace SocketsSample.Hubs
+{
+
```
Comment (davidfowl) [inline samples/SocketsSample/HubCallerContext.cs]: nit: space

File: samples/SocketsSample/IClientProxy.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
+﻿using System.Threading.Tasks;
+
+namespace SocketsSample.Hubs
+{
+
```
Comment (davidfowl) [inline samples/SocketsSample/IClientProxy.cs]: nit: space

File: samples/SocketsSample/IHubConnectionContext.cs [hunk 3]
```diff
@@ -0,0 +1,14 @@
+﻿namespace SocketsSample.Hubs
+{
+
```
Comment (davidfowl) [inline samples/SocketsSample/IHubConnectionContext.cs]: nit: space

File: samples/SocketsSample/EndPoints/HubEndPoint.cs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        {
+            var scopeFactory = _serviceProvider.GetRequiredService<IServiceScopeFactory>();
+
+            try
+            {
+                await _lifetimeManager.OnConnectedAsync(connection);
+
+                using (var scope = scopeFactory.CreateScope())
+                {
+                    var value = scope.ServiceProvider.GetService<THub>() ?? Activator.CreateInstance<THub>();
```
Comment (moozzyk) [inline samples/SocketsSample/EndPoints/HubEndPoint.cs]: nit: rename value to hub?

File: samples/SocketsSample/EndPoints/HubEndPoint.cs [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+                    await value.OnConnectedAsync();
+                }
+
+                await base.OnConnected(connection);
+            }
+            finally
+            {
+                using (var scope = scopeFactory.CreateScope())
+                {
+                    var value = scope.ServiceProvider.GetService<THub>() ?? Activator.CreateInstance<THub>();
```
Comment (moozzyk) [inline samples/SocketsSample/EndPoints/HubEndPoint.cs]: same as above...

File: samples/SocketsSample/EndPoints/HubEndPoint.cs [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        }
+
+        protected override void AfterInvoke(Connection connection, object endpoint)
+        {
+            // Poison the hub make sure it can't be used after invocation
+        }
+
+        protected override void DiscoverEndpoints()
+        {
+            RegisterRPCEndPoint(typeof(THub));
```
Comment (moozzyk) [inline samples/SocketsSample/EndPoints/HubEndPoint.cs]: Should RpcEndPoint be generic as well? The implementation of DiscoverEndpoints there is quite useless.

Comment (davidfowl) [reply samples/SocketsSample/EndPoints/HubEndPoint.cs]: Yea, I ended up doing that.

**Pull request merged**